libdebug/snapshots/process/process_snapshot.py

```diff
@@ -0,0 +1,284 @@
+#
+# This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
+# Copyright (c) 2024 Francesco Panebianco. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for details.
+#
+from __future__ import annotations
+
+import json
+from base64 import b64decode, b64encode
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from libdebug.data.memory_map import MemoryMap
+from libdebug.data.symbol import Symbol
+from libdebug.data.symbol_list import SymbolList
+from libdebug.liblog import liblog
+from libdebug.snapshots.memory.memory_map_snapshot_list import MemoryMapSnapshotList
+from libdebug.snapshots.memory.snapshot_memory_view import SnapshotMemoryView
+from libdebug.snapshots.process.process_shapshot_diff import ProcessSnapshotDiff
+from libdebug.snapshots.registers.snapshot_registers import SnapshotRegisters
+from libdebug.snapshots.snapshot import Snapshot
+from libdebug.snapshots.thread.lw_thread_snapshot import LightweightThreadSnapshot
+
+if TYPE_CHECKING:
+    from libdebug.debugger.internal_debugger import InternalDebugger
+    from libdebug.snapshots.diff import Diff
+
+
+class ProcessSnapshot(Snapshot):
+    """This object represents a snapshot of the target process. It holds information about the process's state.
+
+    Snapshot levels:
+    - base: Registers
+    - writable: Registers, writable memory contents
+    - full: Registers, stack, all readable memory contents
+    """
+
+    def __init__(
+        self: ProcessSnapshot, debugger: InternalDebugger, level: str = "base", name: str | None = None
+    ) -> None:
+        """Creates a new snapshot object for the given process.
+
+        Args:
+            debugger (Debugger): The thread to take a snapshot of.
+            level (str, optional): The level of the snapshot. Defaults to "base".
+            name (str, optional): A name associated to the snapshot. Defaults to None.
+        """
+        # Set id of the snapshot and increment the counter
+        self.snapshot_id = debugger._snapshot_count
+        debugger._snapshot_count += 1
```

Comment:
I'm not sure that this external class should be incrementing a private property of the debugger, but it's fine